cpp/src/arrow/compute/exec/hash_join_benchmark.cc

```diff
@@ -0,0 +1,406 @@
+  BenchmarkSettings settings;
+  settings.num_threads = static_cast<int>(st.range(0));
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_build_batches * 8;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_BuildParallelism(benchmark::State& st) {
+  BenchmarkSettings settings;
+  settings.num_threads = static_cast<int>(st.range(0));
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_threads;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_NullPercentage(benchmark::State& st) {
+  BenchmarkSettings settings;
+  settings.null_percentage = static_cast<double>(st.range(0)) / 100.0;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+std::vector<std::string> keytypes_argnames = {"Hashtable krows"};
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{int32}", {int32()})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{int64}", {int64()})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{int64,int64}", {int64(), int64()})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{int64,int64,int64}",
+                  {int64(), int64(), int64()})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{int64,int64,int64,int64}",
+                  {int64(), int64(), int64(), int64()})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{utf8}", {utf8()})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 64);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{fixed_size_binary(4)}",
+                  {fixed_size_binary(4)})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{fixed_size_binary(8)}",
+                  {fixed_size_binary(8)})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{fixed_size_binary(16)}",
+                  {fixed_size_binary(16)})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{fixed_size_binary(24)}",
+                  {fixed_size_binary(24)})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{fixed_size_binary(32)}",
+                  {fixed_size_binary(32)})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
```

Comment:
Since these benchmarks are all using the same arguments, you could make maintenance easier by factoring them out in a function, e.g.:
```c++
void BM_HashJoinSetArgs(benchmark::internal::Benchmark* bench) {
  bench->ArgNames(...)->RangeMultiplier->(etc.);
}

BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{fixed_size_binary(32)}",
                  {fixed_size_binary(32)})->Apply(BM_HashJoinSetArgs);
```

Reply:
I just factored out the hashtable size out so I don't think this is needed anymore.